src/task.rs

```diff
@@ -150,24 +149,26 @@ unsafe fn set_initial_stack(stack_ptr: *const u32, entry_point: fn(*const u32),
     *cur_ptr = idle as u32;
     cur_ptr = (cur_ptr as u32 - 4) as *mut u32;
 
-    //Set all of the registers
-    //Since we need R0 to be the parameter, and the other registers will be
-    //cleared on use, we'll set them all to the parameter
+    //get the address of the parameter
+    let param_ptr = match param {
+        Some(val) => Box::into_raw(val),
+        None => cur_ptr as *mut T,
+    };
+
     for _i in 0..13 {
-        *cur_ptr = param_val;
+        *cur_ptr = param_ptr as u32;
         cur_ptr = (cur_ptr as u32 - 4) as *mut u32;
```

Comment:
Other than only putting the param in R0, like you said earlier, this PR looks good to me